### ISSUE TITLE
Support for Python based Qt applications

### ISSUE THREAD

**Issue by probonopd** [TheFutureShell/updatedeployqt#10]
Can this also work with Python based Qt applications, such as https://github.com/Ultimaker/Cura/releases/download/4.2-beta/Ultimaker_Cura-4.2.0-BETA.AppImage?

**Comment by antony-jr** [comment]
I used on Pext and it worked but can't tell unless I try it.

**Comment by antony-jr** [comment]
Okay it did not work. Its weird. I will try to debug this.

**Comment by antony-jr** [comment]
One possible cause is mismatch of libqxcb Qt version. I don't know if I can fix this. Why can't we drop language independent and only support native applications? like applications built with c plus plus. I've also tried with Pext and it worked but I can't integrate the option into the menu. Maybe we need to do something else other than modifying the qt xcb plugin which is very elemental. I think we need to inject this into something high level.

**Comment by probonopd** [comment]
I'd really like to see Ultimaker Cura to get a self-updater built in, it's a really popular application.

**Comment by antony-jr** [comment]
I will try but I can't guarantee success.

**Comment by antony-jr** [comment]
Only this particular application is facing this problem. I compared the qxcb version and the pre-built version is just a single patch version old, So I hope this could work if we match the exact version?. I'm gonna try it later. I don't know if this could work.

This is the error I get,

```
QFactoryLoader::QFactoryLoader() looking at "/tmp/.mount_Cura-x51Gm0r/usr/bin/qt/plugins/platforms/libqxcb.so"
Found metadata in lib /tmp/.mount_Cura-x51Gm0r/usr/bin/qt/plugins/platforms/libqxcb.so, metadata=
{
    "IID": "org.qt-project.Qt.QPA.QPlatformIntegrationFactoryInterface.5.3",
    "MetaData": {
        "Keys": [
            "xcb"
        ]
    },
    "className": "QXcbIntegrationPlugin",
    "debug": false,
    "version": 330240
}


Got keys from plugin meta data ("xcb")
QFactoryLoader::QFactoryLoader() checking directory path "/tmp/.mount_Cura-x51Gm0r/usr/bin/platforms" ...
Cannot load library /tmp/.mount_Cura-x51Gm0r/usr/bin/qt/plugins/platforms/libqxcb.so: (/tmp/.mount_Cura-x51Gm0r/usr/bin/libQt5Core.so.5: version `Qt_5.13' not found (required by /usr/lib/libQt5XcbQpa.so.5))
QLibraryPrivate::loadPlugin failed on "/tmp/.mount_Cura-x51Gm0r/usr/bin/qt/plugins/platforms/libqxcb.so" : "Cannot load library /tmp/.mount_Cura-x51Gm0r/usr/bin/qt/plugins/platforms/libqxcb.so: (/tmp/.mount_Cura-x51Gm0r/usr/bin/libQt5Core.so.5: version `Qt_5.13' not found (required by /usr/lib/libQt5XcbQpa.so.5))"
```

It looks like it uses the host system version of libQt5XcbQpa.so.5 even though it exists in the AppImage. When the qtxcb plugin is not replaced, everything just works fine.

**Comment by antony-jr** [comment]
So is this an AppImage error or the plugin. It seems like the plugin fails to load since it can't find libQt5XcbQpa.so.5 (built with Qt 5.10.1) from the system. Since the system will have my version of Qt which is 5.13.0, the load fails. But why this happens when I change the plugin? Maybe they are doing something wrong?

**Comment by antony-jr** [comment]
I tried the same on FreeCAD but the plugin does not do anything? But the application ran without any errors. Python is weird.

**Comment by antony-jr** [comment]
@probonopd Okay I found the problem. Its not with updatedeployqt or AppImages. Ultimate Cura simply packaged the Application wrong.
The plugin started working after I done the following changes...

Change the entry script like so,
```
#!/bin/sh

scriptdir=$(dirname $0)

export PYTHONPATH=$scriptdir/lib/python3.5
#export QT_PLUGIN_PATH=$scriptdir/qt/plugins
#export QML2_IMPORT_PATH=$scriptdir/qt/qml
export QT_QPA_FONTDIR=/usr/share/fonts
#export QT_XKB_CONFIG_ROOT=/usr/share/X11/xkb
export LD_LIBRARY_PATH="$scriptdir"
# Use the openssl.cnf packaged in the AppImage
export OPENSSL_CONF=$scriptdir/openssl.cnf

cura "$@"
```

**Adding LD_LIBRARY_PATH is the key.**

Add **qt.conf**.

```
[Paths]
Prefix = .
Plugins = qt/plugins
Imports = qml
Qml2Imports = qt/qml
```

And now it works. I think the major problem with python programs packaged with AppImages is that they are fundamentally broken in some way. Since we have no standard way of packaging python apps, Its really hard.

**Comment by antony-jr** [comment]
As expected, uses the same theme as the parent application...


![Screenshot_20190801_070524](https://user-images.githubusercontent.com/33930174/62259136-ebd8f080-b42a-11e9-9981-ca1d6ee39be4.png)

**Comment by antony-jr** [comment]
One thing I can't figure out is integrating it into the menu. I can't search by text or id because its qml. I have to try a little harder.

**Comment by antony-jr** [comment]
Also on close the application crashes with 
```
Segmentation fault      (core dumped) cura "$@"
```

Since its on close, it is not noticeable but I think Qt's parent to child de-allocation is the cause.

**Comment by antony-jr** [comment]
Hmm.. FreeCAD now works along with the plugin.

![Screenshot_20190801_075910](https://user-images.githubusercontent.com/33930174/62261304-6c4f1f80-b432-11e9-9b7b-603f4a81c5af.png)


**But the crash at end is still present, I think this is a bug in the plugin.**

**Comment by antony-jr** [comment]
I'm thinking of removing the auto update feature entirely and force integrate with the menu instead.

**Comment by probonopd** [comment]
I think Cura does not have a "Check for updates..." menu entry yet, so you'd need to add one?

**Comment by probonopd** [comment]
Let me know when you think this is ready for being suggested to the Cura team. Thanks :+1:

**Comment by antony-jr** [comment]
> I think Cura does not have a "Check for updates..." menu entry yet, so you'd need to add one?

The problem is that, with current version of code we can't find anything related to menu with an QObject name and that's the problem which needs to be fixed. This problem is encountered only with Python and QML based Qt Applications.

**Comment by antony-jr** [comment]
Searched everywhere, Its impossible to modify QML objects from C++ unless they have some code to actually integrate the C++ code into their QML code.

**At best all I can do is an auto update for any python or QML application based on Qt.** Maybe in the future we can see if this is possible.

But native application which does not use QML seems to work very good. I think this project is gonna hit a dead end.

**Comment by probonopd** [comment]
As an alternative, could the updater called by editing e.g., the Cura Python source code? (How?)

**Comment by probonopd** [comment]
Something along the lines of

```python
if (("APPIMAGE" in os.environ) and ("APPDIR" in os.environ)):
    # We know that we are inside an AppImage
    # Add a menu entry that calls our updater
    ...
else:
    ...
```

**Comment by antony-jr** [comment]
That is the thing, If you use some kind of proxy program(like AppRun) we can't do the integration. Remember that **Qt Plugins** executes inside the program and thus can access all memory of the program. Since we were able to access the memory which is normally not possible due to security reasons, It was possible to manipulate the program. But QML and Python apps seems to do a lot of high level stuff and Its hard to track down which memory corresponds to what. And this is the fundamental problem with this idea. **Its kind of hacking**. Hacks only work for sometime until it gets fixed. So I really don't see any future for this. But One thing we can do is provide a Qt plugin and make the users use it(i.e They have to modify the source file). Since Qt Plugins can work on any platform and programming language. They can get the updater very easily like how **Sparkle framework is used**.

**I will keep on experimenting new stuff but until I find a solution to manipulate  QML and python apps, the development will be stalled.**

**Comment by probonopd** [comment]
Thanks @antony-jr. I believe there is a misunderstanding:

* For compiled apps, it is great to have a way to install the updater without having to change the source code of the application (especially for closed-source applications)
* For Python applications, I think it is _acceptable_ that one has to edit the source code of the application, especially since Python applications are "open source" by default (Python is an interpreted rather than compiled language)

Hence the question is: How to invoke your updater from Python.

**Comment by antony-jr** [comment]
@probonopd Sorry for the delay, I had a lot of work.

Here is the summary,

**For non-QML based Qt Applications[(i.e) Any application coded in any language uses the Qt framework or bindings to Qt framework(like in python)], can be easily integrated by the update tool with no change to the source files.**

Now for **QML based Qt Application**, regardless of the programming language(even C++), you cannot integrate the updater without modifying the QML sources. **I like to reiterate, you have to modify the QML source but not the C++ or other language which loads the QML source.**

What that means is that for **QML based Qt Application** we can run a automatic check to search for the menu item written in the QML file(s) or the developer has to give the main qml file which contains the menu bar. In case of **Cura** its here -> https://github.com/Ultimaker/Cura/blob/f2044998ae360b9fa195d497f3c23c92265dce1a/resources/qml/MainWindow/ApplicationMenu.qml#L96

It was really straight forward with **github search** but we would have to implement it into our deployer.

After finding the required QML source, we add the code to start the plugin from QML itself. .We don't have to load the plugin from C++ at all. So for QML apps its really straight forward. Copy the plugin to ```QMLDIR``` and import it. **This means we don't need to override any Qt plugins, what that  means is that QML apps have a great future.**

TL;DR

**QML based Qt Apps(Any Language)**: copy the plugin, and modify QML source only, no need to re-compile.

**Non QML based Qt Apps(Any Language)**: override qt plugin then copy the plugin, no need to modify source. No need to re-compile.

**For both cases we don't need to re-compile at all.**

Does that sound good?

**Comment by probonopd** [comment]
Sounds excellent, sounds it could work for Cura?
